lib/tracer/tracer.js

```diff
@@ -89,62 +108,122 @@ Tracer.prototype.event = function (traceInfo, type, data, metaData) {
   }
 
   if (data) {
-    let info = _.pick(traceInfo, 'type', 'name');
+    let info = _.pick(trace, 'type', 'name');
     event.data = this._applyFilters(type, data, info, 'start');
   }
 
-  if (metaData && metaData.name) {
-    event.name = metaData.name;
-  }
-
   if (Kadira.options.eventStackTrace) {
     event.stack = CreateUserStack();
   }
 
-  eventLogger('%s %s', type, traceInfo._id);
+  eventLogger('%s %s', type, trace._id);
 
-  if (lastEvent && !lastEvent.endAt) {
-    if (!lastEvent.nested) {
-      console.error('Monti: invalid trace. Please share the trace below at');
-      console.error('Monti: https://github.com/monti-apm/monti-apm-agent/issues/14');
-      console.dir(traceInfo, { depth: 10 });
-    }
-    let lastNested = lastEvent.nested[lastEvent.nested.length - 1];
-
-    // Only nest one level
-    if (!lastNested || lastNested.endAt) {
-      lastEvent.nested.push(event);
-      return event;
-    }
+  event.level = activeEvent ? activeEvent.level + 1 : 0;
 
+  if (event.level > 1) {
     return false;
   }
 
-  traceInfo.events.push(event);
+  if (activeEvent && !activeEvent.endAt) {
+    checkForNested(activeEvent);
+    activeEvent.nested.push(event);
+    return event;
+  }
+
+  trace.events.push(event);
 
   return event;
 };
 
-Tracer.prototype.eventEnd = function (traceInfo, event, data) {
+Tracer.prototype.runWithEvent = function (fn, event) {
+  const als = MontiAsyncStorage;
+
+  const partial = {
+    activeEvent: event
+  };
+
+  const prev = als.getStore();
+
+  if (!prev) return fn.call(this, event);
+
+  als.enterWith(Object.assign({}, prev, partial));
+
+  try {
+    return fn.call(this, event);
+  } finally {
+    als.enterWith(prev);
+  }
+};
+
+Tracer.prototype.asyncEvent = function (type, data, meta, fn) {
+  const { info } = getStore();
+
+  if (!info) {
+    return Reflect.apply(fn, this, [null]);
+  }
+
+  const event = this.event(info.trace, type, data, meta);
+
+  return this.runWithEvent(() => {
+    try {
+      let result = Reflect.apply(fn, this, [event]);
+
+      if (result instanceof Promise) {
```

Comment:
We want to instead check if it is a thenable (an object with a `then` method). This is consistent with what js engines and Meteor do.